[MQ4] Implement overflow-block/overflow-inline media query features

nt1m opened the pull request:
#### 3f9d91e4ed1d08b947d964a91254b65e2046ab08
<pre>
[MQ4] Implement overflow-block/overflow-inline media query features
<a href="https://bugs.webkit.org/show_bug.cgi?id=253662">https://bugs.webkit.org/show_bug.cgi?id=253662</a>
rdar://106511968

Reviewed by Darin Adler.

<a href="https://drafts.csswg.org/mediaqueries-4/#mf-overflow-block">https://drafts.csswg.org/mediaqueries-4/#mf-overflow-block</a>
<a href="https://drafts.csswg.org/mediaqueries-4/#mf-overflow-inline">https://drafts.csswg.org/mediaqueries-4/#mf-overflow-inline</a>

* LayoutTests/fast/css/media-query-overflow-block-dynamic-expected.html: Added.
* LayoutTests/fast/css/media-query-overflow-block-dynamic.html: Added.
* LayoutTests/fast/css/media-query-overflow-block-paged-expected.html: Added.
* LayoutTests/fast/css/media-query-overflow-block-paged.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/mediaqueries/test_media_queries-expected.txt:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/query/MediaQueryFeatures.cpp:
(WebCore::MQ::Features::overflowBlock):
(WebCore::MQ::Features::overflowInline):
* Source/WebCore/css/query/MediaQueryFeatures.h:

Canonical link: <a href="https://commits.webkit.org/263088@main">https://commits.webkit.org/263088@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81fd03331c520314a15d8a7c229e4ebb647118b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3569 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3755 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4993 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3843 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3548 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3709 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3657 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3104 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3612 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3226 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4815 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/1374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3199 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/3219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3175 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3258 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/4596 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3626 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2929 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3199 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/863 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3194 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3445 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/mediaqueries/test_media_queries-expected.txt:
```diff
@@ -1256,20 +1256,20 @@ PASS should_not_apply: all and min-color : 1
+PASS expression_should_be_known: overflow-block: none
+PASS expression_should_be_known: overflow-block: paged
+PASS expression_should_be_known: overflow-block: scroll
 FAIL expression_should_be_known: overflow-block: optional-paged assert_true: expected true got false
```
optional-paged was removed from the spec and this WPT, next import should make this pass.

darinadler commented on Source/WebCore/css/query/MediaQueryFeatures.cpp:
```diff
@@ -689,6 +689,40 @@ const FeatureSchema& displayMode()
+        "overflow-block"_s,
+        Vector { CSSValueNone, CSSValueScroll, CSSValuePaged },
+        [](auto& context) {
+            // FIXME: Implement none.
```
Is this too complex to include in the same patch?

nt1m replied:
It's non trivial, `none` should only match when a WKWebView disables scrollbars using a AppKit preference, which is quite rare. Unfortunately, this data isn't available to WebCore easily, so `scroll` will match instead.

darinadler commented on Source/WebCore/css/query/MediaQueryFeatures.cpp:
```diff
@@ -689,6 +689,40 @@ const FeatureSchema& displayMode()
+        [](auto& context) {
+            // FIXME: Implement none.
+            bool usesPaginatedMode = [&] {
+                auto& frame = *context.document.frame();
```
What guarantees `frame` is non-null?

nt1m replied:
This assumption seems to be done all over this file: https://searchfox.org/wubkat/search?q=context.document.frame%28%29&path=&case=false&regexp=false

darinadler commented on Source/WebCore/css/query/MediaQueryFeatures.cpp:
```diff
@@ -689,6 +689,40 @@ const FeatureSchema& displayMode()
+                auto* frameView = frame.view();
+                if (!frameView)
+                    return false;
+                return frameView->pagination().mode != Pagination::Unpaginated;
```
I like to use &&:
```
return frameView && frameView->pagination().mode != Pagination::Unpaginated;
```

darinadler commented on Source/WebCore/css/query/MediaQueryFeatures.cpp:
```diff
@@ -689,6 +689,40 @@ const FeatureSchema& displayMode()
+{
+    static MainThreadNeverDestroyed<IdentifierSchema> schema {
+        "overflow-block"_s,
+        Vector { CSSValueNone, CSSValueScroll, CSSValuePaged },
```
`Vector` is not the ideal data structure for something we do not intend to resize. Maybe `std::initializer_list<CSSValueID>` is what `FeatureSchema` should use, or maybe `FixedVector`.

Related: Do we actually have to write `Vector` here or can we just omit the class template name? If we can omit it, that makes future refactoring easier.

nt1m replied:
I can try omitting `Vector` and refactoring `FeatureSchema` away from `Vector` in a separate patch.

nt1m replied:
If I omit `Vector` the compiler can't automatically infer the type of the schema

darinadler commented on Source/WebCore/css/query/MediaQueryFeatures.cpp:
```diff
@@ -689,6 +689,40 @@ const FeatureSchema& displayMode()
+        Vector { CSSValueNone, CSSValueScroll },
+        [](auto& context) {
+            // FIXME: Implement none.
+            UNUSED_PARAM(context);
```
I prefer to omit the argument name rather than using `UNUSED_PARAM` when it’s practical to do so.

nt1m committed:
[MQ4] Implement overflow-block/overflow-inline media query features
https://bugs.webkit.org/show_bug.cgi?id=253662
rdar://106511968

Reviewed by Darin Adler.

https://drafts.csswg.org/mediaqueries-4/#mf-overflow-block
https://drafts.csswg.org/mediaqueries-4/#mf-overflow-inline

* LayoutTests/fast/css/media-query-overflow-block-dynamic-expected.html: Added.
* LayoutTests/fast/css/media-query-overflow-block-dynamic.html: Added.
* LayoutTests/fast/css/media-query-overflow-block-paged-expected.html: Added.
* LayoutTests/fast/css/media-query-overflow-block-paged.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/mediaqueries/test_media_queries-expected.txt:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/query/MediaQueryFeatures.cpp:
(WebCore::MQ::Features::overflowBlock):
(WebCore::MQ::Features::overflowInline):
* Source/WebCore/css/query/MediaQueryFeatures.h:

Canonical link: https://commits.webkit.org/263088@main
